examples/src/main/scala/com/github/niqdev/ToUpperCaseApp.scala

```diff
@@ -22,7 +22,7 @@ object ToUpperCaseTopology {
       topology <- ZStreamsBuilder { builder =>
         for {
           sourceStream <- builder.stream[String, String](sourceTopic)
-          sinkStream   <- sourceStream.mapValue(_.toUpperCase)
+          sinkStream   <- sourceStream.mapValue(_.length)
```

Comment:
I think this is an interesting case...
```scala
(sinkStream: ZKStream[String, Int])   <- sourceStream.mapValue(_.length)
_                                     <- sinkStream.to(sinkTopic)
```
the *output type* of the topic is purely determined by the types of the `ZKStream[K, V]` wrapper, you can't define for instance
```scala
def to[K0, VO](topic: String): RIO[KafkaStreamsConfig, Unit]
```
because that's how the scala/java KStream API works

So this will always compile until there is an instance for `Codec[T]` in scope, which [exists](https://github.com/niqdev/zio-kafka-streams/blob/master/modules/serde/src/main/scala/kafka/streams/serde/Codec.scala#L36) for all basic types

On the other hand, I believe in most of the cases you would use Avro, and in this case this shouldn't be a problem at all, because you need to defined an instance of `AvroCodec` for your type `T` and that won't compile unless the types are what you expect

```scala
// here dummy.value.length doesn't compile
(sinkStream: ZKStream[DummyKey, DummyValue])   <- sourceStream.mapValue(dummy => DummyValue(dummy.value.toUpperCase))
_                                              <- sinkStream.toAvro(sinkTopic)
```

wdyt?

Reply:
In a greenfield project I agree avro is the state of the art and should be used. But there are many real case scenarios in which this is not possible or preferable. Smaller projects, brownfield projects that already have topics without avro (true story), systems that use another schema software or different binary encodings...

Reply:
fair enough, but at this point, the issue is that there are default codecs already defined and we are using basic types... it's probably still not ideal and won't solve the issue completely, but if you have your own type, you can build your own codec using `Codec[T].cmap`, similar to
```scala
implicit val uuidCodec: Codec[java.util.UUID] =
  stringCodec.cmap(...)
```
and catch most of these errors at compile time

Reply:
I don't think I totally understood what you are saying here. Are you saying that we could make it type safe by using slightly different codecs (for example `UUID` instead of `String`)?

Reply:
> it's probably still not ideal and won't solve the issue completely

the current implementation by default uses implicit codecs and has the types "dictated" by the ZKStream constructor and you won't get around it easily